### ISSUE TITLE
[RFC] Register model from Cloud providers buckets

### ISSUE THREAD

**Issue by khoaisohd** [opensearch-project/ml-commons#1926]
## Problem statement
Today, customers can register models to OpenSearch using public URLs or local file systems. There is no solution to register ML models from Cloud providers. The only way is to make bucket objects public and use the public URL, therefore exposing ML models to anyone. This makes the process of register a model quite tedious as we need to upload the model in chunks from our local because we do not want to expose ML models to the public internet.

## Scope
Developers can register models from the Cloud providers seamlessly without exposing to public internet 

### Option 1
To allow customers to upload models from cloud providers to OpenSearch, we need to extend the model input to allow customers to provide information about the cloud provider and the model file location. Furthermore, cloud provider Java SDK dependency (OCI SDK, AWS SDK) is also added to the ml-commons plugin to allow it to download files from cloud provider storage service

When the customer uploads a model from the cloud provider, here is what ml-common plugins do

- Look at the scheme of the URI and determine the cloud provider storage service.
   -  **S3 format**: s3://{bucket}/{key} or s3://{access-point-arn}/{key} 
   -  **GCS format**: gs://{bucket}/{model-file}
   -  **OCI object storage format**: oci-os://{namespace}/{bucket}/{modile-file}
- Add more settings from the model input for AuthN and AuthZ, so the settings will be prefixed by the cloud provider name for example 
   - aws_access_key
   - aws_secret_key
   - aws_session_token
   - aws_service_name
   - aws_region 
- Download model from the cloud provider using cloud provider SDK

Customer experience for using S3
```
{
    "name": "phuong6",
    "model_group_id": "jj1nh4sBNGgnXkAx2296",
    "version": "1.0.0",
    "description": "BYOM Model",
    "model_format": "TORCH_SCRIPT",
    "function_name": "SPARSE_ENCODING",
    "model_content_hash_value": "e13b74006290a9d0f58c1376f9629d4ebc05a0f9385f40db837452b167ae9021",
    "url":  "s3://phuong_bucket/traced_small_model.zip",
    "aws_access_key": "{your-aws-access-key}",
    "aws_secret_key": "{your-aws-secret-key}",
    "aws_session_token": "{your-aws-session-token}",
    "aws_service_name": "{your-aws-service-name}"
    "aws_region": "us-east-1"
}
```
Pros
- Isolated changes can be added to [ModelHelper](https://github.com/opensearch-project/ml-commons/blob/main/ml-algorithms/src/main/java/org/opensearch/ml/engine/ModelHelper.java) without impacting other component

Cons
- Additional complexity to handle confidential information like access key, secret key. We need to reimplement the same logic for encrypting or decrypting credentials in connector component
- Additional complexity in the future if we want to connect to raw HTTP with customer provided credentials, most likely we need to re-implement the logic for templating
- Whenever we need to support a new cloud provider or update existing cloud provide, we might need to make changes in 2 places MLRegisterModelInput and connector

### Option 2
Leverage on existing component [Connector](https://github.com/opensearch-project/ml-commons/blob/main/common/src/main/java/org/opensearch/ml/common/connector/Connector.java)
- Add a new action type **DOWNLOAD**
- Expose another function in [RemoteConnectorExecutor](https://github.com/opensearch-project/ml-commons/blob/d0895bb1f361d1a8f0f0b897abea89ad9dcdb0f0/ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/RemoteConnectorExecutor.java#L33) to execute download action as stream
```
default InputStream executeDownload() {
}
```
- Add **url_connector** in MLReigsterModelInput
- Update [ModelHelper](https://github.com/opensearch-project/ml-commons/blob/main/ml-algorithms/src/main/java/org/opensearch/ml/engine/ModelHelper.java) to trigger download model through connector

Customer experience 
```
POST /_plugins/_ml/models/_upload
{
    "name": "phuong11",
    "version": "1.0.0",
    "description": "Sample to load model from cloud provider",
    "model_format": "TORCH_SCRIPT",
    "function_name": "TEXT_EMBEDDING",
    "model_content_hash_value": "f8012a4e6b5da1f556221a12160d080157039f077ab85a5f6b467a47247aad49",
    "url_connector": {
        "protocol": "aws_sigv4",
        "credentials": {
            "access_key": "secret_access_key",
           "secret_key": "secret_key..."
        },
        "actions": [
            {
                "method": "GET",
                "action_type": "DOWNLOAD",
                "url": "https://my-s3-bucket.s3.amazonaws.com/example.txt"
            }
        ]
    }
}
```

Pros
- Less effort
- Support not only cloud provider but also HTTP protocol that requires credentials
- Unify all the logic to talk to cloud provide in one place [Connector](https://github.com/opensearch-project/ml-commons/blob/main/common/src/main/java/org/opensearch/ml/common/connector/Connector.java)

Cons
- Connector is design to invoke remote model functionality, in this case there is no existing remote model,  we just download it through connector, not sure it will be align with the long term plan for connector component
- Some refactoring in [RemoteConnectorExecutor](https://github.com/opensearch-project/ml-commons/blob/d0895bb1f361d1a8f0f0b897abea89ad9dcdb0f0/ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/RemoteConnectorExecutor.java#L33) to avoid code duplication between executePredict and executeDownload and it might impact the predict feature

**Comment by HenryL27** [comment]
I think going the connector route makes the most sense to me. Yes, it's currently designed exclusively for doing external model inferences, but at its heart it's just a primitive for talking to things that aren't OpenSearch. It shouldn't be too too difficult to use for model downloads, right? 
I almost wonder if it's worth moving connectors into core. Then we could do things like remote procedure calls or external DB lookups or whatever.

**Comment by khoaisohd** [comment]
Yeah, it is very true. I'm adding support for OCI cloud providers in multiple plugins and it turns out that different plugin (for example notification plugin) has different ways to call cloud provider services and passing credentials. If connector can be moved core then it will make it much easier to support new type of cloud provider. Furthermore, it also make user experience consistent across all plugins, since they can just have a same way to provide credential and call cloud provider service APIs

**Comment by HenryL27** [comment]
Oh, how do they do it in notifications?

**Comment by khoaisohd** [comment]
Here is for [notification plugin](https://opensearch.org/docs/latest/observing-your-data/notifications/api/), there is no credentials

```
"sns": {
  "topic_arn": "<arn>",
  "role_arn": "<arn>" //optional
}
"slack": {
  "url": "https://sample-chime-webhoook"
}
"chime": {
  "url": "https://sample-amazon-chime-webhoook"
}
"webhook": {
      "url": "https://custom-webhook-test-url.com:8888/test-path?params1=value1&params2=value2"
}
"smtp_account": {
  "host": "test-host.com",
  "port": 123,
  "method": "start_tls",
  "from_address": "test@email.com"
}
"ses_account": {
  "region": "us-east-1",
  "role_arn": "arn:aws:iam::012345678912:role/NotificationsSESRole",
  "from_address": "test@email.com"
}
"email_group": { //Email recipient group
  "recipient_list": [
    {
      "recipient": "test-email1@test.com"
    },
    {
      "recipient": "test-email2@test.com"
    }
  ]
}
"email": { //The channel that sends emails
  "email_account_id": "<smtp or ses account config id>",
  "recipient_list": [
    {
      "recipient": "custom.email@test.com"
    }
  ],
  "email_group_id_list": []
}
```

**Comment by samuel-oci** [comment]
thank you for the RFC @khoaisohd , I agree with @HenryL27 and I tend to favor the connector part better because it exposes a unified interface more consistent with interacting externally.
Regarding moving connectors to core and other plugins, it's an interesting suggestion, if going with the connector is accepted for ml-commons we should probably use it as a show case example in other plugins as well and bring about an additional RFC.

**Comment by khoaisohd** [comment]
Yeah, I'm also more inclined to the connector especially the option 3 since it is going to give us a lot of flexibility

**Comment by ylwu-amzn** [comment]
Thanks @khoaisohd, this is a good proposal. 

Our original design for connector considers a lot for extensible.
1. Support more protocols. We are not just focusing on HTTP. Connector have `protocol` field for future extension to support other protocols like FTP, gRPC etc.
2. Support more action type. Action type should be general, like predict for general remote inference purpose. We also considered adding train action type for flowing data from OpenSearch to a remote training service. 
3. Connector could have multiple actions to do a more complicated task, so connector may contain a list of actions.

So the option2 is aligned with our original extension support design. 

The connector is to make the interface easier, flexible/extensible to connect to any service with any protocol. Option3 looks like a more basic functionality or some lower layer. I think it's good if OpenSearch core can provide this. We did't build connector like this way to reduce the interface complexity. I think we may still need connector layer with option3. 

>  I'm adding support for OCI cloud providers in multiple plugins and it turns out that different plugin (for example notification plugin) has different ways to call cloud provider services and passing credentials. If connector can be moved core then it will make it much easier to support new type of cloud provider. Furthermore, it also make user experience consistent across all plugins, since they can just have a same way to provide credential and call cloud provider service APIs.

Thanks for your like of connector framework. I think moving to OpenSearch core needs more discussion. Our current design is providing a client jar which can be consumed by other plugins. 

For example, neural-search plugin is using it today https://github.com/opensearch-project/neural-search/blob/main/build.gradle#L237

We are going to release a new plugin flow-framework in 2.12 which also depends on ml-commons, you can check how they use ML client to create connector https://github.com/opensearch-project/flow-framework/blob/main/src/main/java/org/opensearch/flowframework/workflow/CreateConnectorStep.java#L172

**Comment by khoaisohd** [comment]
Greate, I'll go ahead option 2

**Comment by Zhangxunmt** [comment]
@khoaisohd Thanks for this proposal. With option2, one concern is the security risks. Currently we only allow uploading pre-trained model provided by open search due to security reasons. Similarly, we need to be careful for how we handle security risk by allowing customer to download a model from his own cloud. A lot of limitations may need to apply based on the past security review.

**Comment by khoaisohd** [comment]
Hi @Zhangxunmt, can you explain more about security risks, and how is it mitigated by the current process
* Using public URL for downloading models
* Using local file (after customer upload the models files as local file)

I will reuse the same protection mechanism for download a model from cloud provider.

**Comment by Zhangxunmt** [comment]
@khoaisohd, in AOS, it's recommended to use remote model as the first choice, and provided pre-trained model as the second. If someone really needs to upload his model from public URL, security team require the model to be reviewed and uploaded by us manually. Basically the limitation is very strict in AOS. 
In Open Source, you can upload from your own url with fewer limitation but some checks are still needed like model size, url legitimacy, etc.

**Comment by khoaisohd** [comment]
@Zhangxunmt I got it so it is pretty strict for AOS, but for this situation, customers want to use their own model, does customer need to work with security team to scan their model before using them. And if they upload to S3 and use a pre-authenticated endpoint without doing security scanning of the model,  how does AOS prevent customers from not using those models through S3 pre-endpoint?

**Comment by samuel-oci** [comment]
> @khoaisohd, in AOS, it's recommended to use remote model as the first choice, and provided pre-trained model as the second. If someone really needs to upload his model from public URL, security team require the model to be reviewed and uploaded by us manually. Basically the limitation is very strict in AOS. In Open Source, you can upload from your own url with fewer limitation but some checks are still needed like model size, url legitimacy, etc.

@Zhangxunmt I am not sure I follow the logic here and how it's relevant for this change. OpenSearch is an open source project designed to run on any platform/vendor. It can be hosted on-premise, or on any cloud provider. As long as the platform is flexible enough to define security policy to do what you need I believe this should solve the issues on the hosted service as well.
For example if customer is using public URL which is allowed today on ml-commons and you wish to block it on your deployment you have the security settings to do so (e.g. `plugins.ml_commons.allow_registering_model_via_url`).

I think if you're concerned with object storage access by connector, the right approach is to discuss the security settings that can block it or limit it in case the admin/service hosting entity (AWS/OCI/Aiven etc..)  require so.

**Comment by Zhangxunmt** [comment]
> @Zhangxunmt I got it so it is pretty strict for AOS, but for this situation, customers want to use their own model, does customer need to work with security team to scan their model before using them. And if they upload to S3 and use a pre-authenticated endpoint without doing security scanning of the model, how does AOS prevent customers from not using those models through S3 pre-endpoint?

@khoaisohd, this is openSource community so let's only focus on the open source world. The security policies are different based on different cloud platform which you need to comply with when you deploy to them.

**Comment by arjunkumargiri** [comment]
@khoaisohd  let's ensure that this feature is behind a feature flag. Ability to download from any data source has multiple security vulnerabilities and users upgrading to latest OpenSearch version might expose download feature without their knowledge. Feature flag enables OpenSearch administrators to control the features that are exposed and enable the feature only when required.

**Comment by samuel-oci** [comment]
> @khoaisohd let's ensure that this feature is behind a feature flag. Ability to download from any data source has multiple security vulnerabilities and users upgrading to latest OpenSearch version might expose download feature without their knowledge. Feature flag enables OpenSearch administrators to control the features that are exposed and enable the feature only when required.

@arjunkumargiri can you provide some clarifications on the process of when something goes under feature flag and when it's coming out of it?
My understanding based on observation so far is that we would put something under a feature flag when it's experimental or the design is incomplete.
For example like we did with the early versions of searchable snapshots before caching was implemented.
But in both cases we had a clear roadmap how to get those out from feature flag when they were released to the version branch.
I think if we plan to put it under feature flag we should also need to be very crisp around the follow up steps to take it out of experimental state. In that case I believe a useful feedback would be:
1. What specific changes we would like to see in this feature to consider the design complete?
2. The process around taking something out of a feature/experimental flag

**Comment by arjunkumargiri** [comment]
@samuel-oci feature flag will be used as plugin configuration rather than as an experimental feature. Similar to setting `plugins.ml_commons.allow_registering_model_via_url` to support download action type through connectors let's add a settings `plugins.ml_commons.allow_download_action_type` with default value as `false`. This setting enables opensearch administrators to  decide if the feature needs to be turned on/off.

**Comment by mingshl** [comment]
@khoaisohd would you please provide some updates on this RFC? I see this ticket is in progress. Would you like to also assign this issue to yourself ?

**Comment by nathaliellenaa** [comment]
@khoaisohd Are you still working on this? If not, feel free for other contributors to work on this feature
